### DISCUSSION TITLE
Fastify as the backend, and a complete rewrite

### DISCUSSION THREAD

**Discussion by janaSunrise** [rslay/ZeroChat#45] [Ideas]
I propose switching to fastify for our further rewrite.

### Why we want to

- Less dependencies, more lightweight
- Better performance, and speed
- Async support
- TypeScript support inbuilt
- Really well written documentation
- Easy to get started and use!

### Changing the project file structure.

I have a better structure in mind, that looks like this
```
app.js
app/
    models/
    classes/
    utils/
    routes/
    public/
   views/
bin/
```

Add a `config.ini` , which will look like this:

```ini
[APP]
MOTD=
ROOMS=$1, $2, $3

[HTTP]
HTTP_PORT=80
HTTPS_PORT=443
USE_HTTPS=false

[CHAT]
DEFAULT_THEME=dark
DEFAULT_ROOM=hallway
DEFAULT_INLINE_PREVIEW=true
```

Rest of the config will go in `.env`, and maybe get the Password authentication working, or Here's a better advice, Instead of having fixed number of rooms, in the config / `rooms.txt`, Maybe allow users to create their room, with a password, and add public + private key authentication. I'm thinking about key verification, and working, I'll notify as it's done.

**Comment by rslay** [comment]
## The idea of not having a "fixed list of rooms"

This comes from a misunderstanding of how zerochat works, there's actually an infinite amount of rooms that exist. It's just that if you click on the Rooms text box, you'll get suggestions, which are in the rooms list.

But you can join **any room** whatsoever, the idea is you can go to any `/suburl` in zerochat, and that'll work as a room.
Try it, go to https://chat.justhack.in/horseRiding, and you'll notice the room field is automatically set to `/horseRiding`
It's kinda like `#channelName` in IRC chats, if you recognize that.

Having a `ROOMS` variable that holds the list of public rooms seems clunky, **since if you want like 20 then that line gets long.**

## Having public/private key pairs

I would love that and I even thought of it already, but the issue is, how do you encrypt a room's text on the fly within the browser if you arent using javascript? The best we can do is use the latest TLS standard and encrypt the site's traffic with a free HTTPS certificate, which is already being done. Maybe a guide on how to do that would be helpful

**Reply by janaSunrise** [reply]
+1 Agreed, then I guess it's solved, Maybe we need to bring password authentication into action, and a note on the `README.md` about the infinite number of rooms, or maybe in the rewrite, we could add a box where they type the room name, and get redirected into that specific room, hence no need to manually change URL, POV of a normal person.

We could also rewrite according to the new structure, and maybe use TS? Fastify supports that too, you just write same code, but it works for both magically!

**Reply by rslay** [reply]
You're totally right, I realize I didn't mention room functionality at all and it isn't exactly an obvious thing.

The point of supporting `/roomName` is so that you can send a link to a friend of a room, when you click "Enter" the URL does already revert to https://chat.justhack.in but the form submitted a POST request which has the user, tripcode pass, and room.

And yes, I'm all for using typescript.

About passwords - this is another thing I hadn't explained, but unless you've used an imageboard, you're probably unfamiliar with the idea of tripcodes. It's a stateless way to prove your identity, without password storage.

I should write this out in the README as well, I'll do that now

**Reply by janaSunrise** [reply]
Great then! So when do we start the migration? I'll make a PR, and we might collab and change everything, basically convert to a new framework, and use Typescript, obv better.

**Reply by rslay** [reply]
I'm going to be busy during the week, but on Friday night for me (Saturday morning for you) I can stay up and work with you on this, especially if you get up a lil bit earlier

**Reply by janaSunrise** [reply]
Great! You can count on me, but I'm not sure about waking up early 😂 I usually wake at 8:30, What do you think about friday, and the weekend too?

**Reply by rslay** [reply]
Yeah those are the best times that work for me too
